src/txn/txn.c

```diff
@@ -1677,9 +1678,17 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
     if (!prepare)
         F_CLR(txn, WT_TXN_TS_ROUND_PREPARED);
 
+    /* Enter the commit generation before timestamp validity check. */
+    __wt_session_gen_enter(session, WT_GEN_TXN_COMMIT);
```

Comment:
Did you think about whether we could skip this check for prepared transactions? They have already given up their transaction ID, and can take a long time to resolve. So ideally we would not wait for their resolution if we didn't have to.

Reply:
Yes, I thought about the prepared transactions as well but if we skip this check for prepared transactions then I believe we could still have a race where a checkpoint can fail to include the prepared transaction updates.

Reply:
If this check needs to be in place for prepared transactions, we definitely can't hold the schema lock while waiting for it to complete during checkpoint.

Could you please describe more clearly the scenario where a prepared transaction would fail? The durable timestamp exists to ensure we know which checkpoint a prepared transaction is included in. Since the checkpoint has already chosen a timestamp at this point, it should never include a prepared transaction that is unresolved.

Reply:
@agorrod This depends on whether we allow to move the stable timestamps beyond the commit timestamp of the prepared timestamp. This should not be possible in MongoDB as I understand it because they syncronize the prepared commit with stable timestamp movement. But it is theoretically still possible to happen. Here's what may happen:

1. Checkpoint starts and context switch before it gets the checkpoint timestamp.
2. Prepare transaction starts to commit and we pass the timestamp checks and context switch.
3. Stable timestamp is moved beyond the prepare transaction's commit timestamp.
4. Checkpoint set the checkpoint timestamp as the new stable timestamp.
5. Checkpoint finishes without the prepared transaction finishes committing.

Reply:
@quchenhao WT raise a PANIC error at the end of the commit transaction if the prepared transaction durable timestamp is less than or equal to the stable timestamp. Still, it is possible that the checkpoint may write some of the prepared transaction data to disk and corrupt the on-disk data. 

The checkpoint corrupting the on-disk data can be fixed with the project of adding transaction id to the committing prepared transaction.

Reply:
Thanks @kommiharibabu for pitching in. Yes, I found a check [here](https://github.com/wiredtiger/wiredtiger/blob/develop/src/txn/txn.c#L1963) in txn_commit to check for the stable timestamp against the prepared transaction's durable timestamp.

Reply:
@agorrod I will make the necessary changes to skip entering into commit generation for the prepared transactions. I have put up a suggestion [here](https://github.com/wiredtiger/wiredtiger/pull/10193#discussion_r1485796165) regarding blocking the schema lock in the `__checkpoint_prepare ` .

Please let me know your thoughts.

Reply:
OK, thanks.
FWIW: I think the stable timestamp can move forward, but the checkpoint timestamp should not be able to move forward (or past it).